### PR TITLE
Make count, figures, revision, and index warmup use non-blocking communication

### PR DESCRIPTION
**Pull request by danielhlarkin**
### Scope & Purpose

This PR changes the behavior of a few collection methods and the associated rest handler to use non-blocking methods for cluster-internal communication, and free up resources for other requests. In particular, it affects counts, figures, collection revision, and the index warmup procedure.

- [x] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [x] The behavior change can be verified via automatic tests

Enterprise component: https://github.com/arangodb/enterprise/pull/323

### Testing & Verification

This change is already covered by existing tests.

Jenkins: http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/6336/

### CODE REVIEW

File: arangod/Cluster/ClusterMethods.cpp [hunk 1]
```diff
@@ -753,211 +850,208 @@ bool smartJoinAttributeChanged(LogicalCollection const& collection, VPackSlice c
...
-
-          if (r.isString()) {
-            VPackValueLength len;
-            char const* p = r.getString(len);
-            TRI_voc_rid_t cmp = TRI_StringToRid(p, len, false);
-
-            if (cmp != UINT64_MAX && cmp > rid) {
-              // get the maximum value
-              rid = cmp;
+  auto cb = [rid](std::vector<Try<network::Response>>&& results) -> std::pair<OperationResult, TRI_voc_rid_t> {
```
Comment (graetzer) [inline arangod/Cluster/ClusterMethods.cpp]: this construct seems weird

File: arangod/Cluster/ClusterMethods.cpp [hunk 2]
```diff
@@ -242,6 +242,103 @@ static void addTransactionHeaderForShard(transaction::Methods const& trx,
...
+          break;
+        }
+      }
+    }
+  }
+  return OperationResult(result);
+}
+
+template <typename T>
+std::pair<OperationResult, T> handleResponsesFromAllShards(
```
Comment (graetzer) [inline arangod/Cluster/ClusterMethods.cpp]: I don't think you necessarily need this, the above version should be able to do the same thing ?

File: arangod/Cluster/ClusterMethods.cpp [hunk 3]
```diff
@@ -753,211 +850,208 @@ bool smartJoinAttributeChanged(LogicalCollection const& collection, VPackSlice c
...
-int warmupOnCoordinator(std::string const& dbname, std::string const& cid) {
+futures::Future<OperationResult> warmupOnCoordinator(std::string const& dbname,
+                                                     std::string const& cid) {
   // Set a few variables needed for our work:
   ClusterInfo* ci = ClusterInfo::instance();
   auto cc = ClusterComm::instance();
   if (cc == nullptr) {
     // nullptr happens only during controlled shutdown
-    return TRI_ERROR_SHUTTING_DOWN;
+    return futures::makeFuture(OperationResult(TRI_ERROR_SHUTTING_DOWN));
```
Comment (graetzer) [inline arangod/Cluster/ClusterMethods.cpp]: I don't think you necessarily need to return a OperationResult. A Result object is also fine no ?

File: arangod/Cluster/ClusterMethods.cpp (hunk 2 above)
Comment (danielhlarkin) [reply arangod/Cluster/ClusterMethods.cpp]: This one allows you to return additional data instead of just the `OperationResult`. Otherwise you need to play some games with passing e.g. a dynamically allocated piece of data via `shared_ptr` capture. Basically this allows for more of a reduce-style handler. Does that make sense?

File: arangod/Cluster/ClusterMethods.cpp [hunk 4]
```diff
@@ -753,211 +850,208 @@ bool smartJoinAttributeChanged(LogicalCollection const& collection, VPackSlice c
...
 
 ////////////////////////////////////////////////////////////////////////////////
 /// @brief returns figures for a sharded collection
 ////////////////////////////////////////////////////////////////////////////////
 
-int figuresOnCoordinator(std::string const& dbname, std::string const& collname,
-                         std::shared_ptr<arangodb::velocypack::Builder>& result) {
+futures::Future<OperationResult> figuresOnCoordinator(
+    std::string const& dbname, std::string const& collname,
+    std::shared_ptr<arangodb::velocypack::Builder> result) {
```
Comment (graetzer) [inline arangod/Cluster/ClusterMethods.cpp]: couldn't you put the figures into the OperationResult ? An extra return parameter outside of the Future feels like it doesn not quite fit the concept

File: arangod/Cluster/ClusterMethods.cpp (hunk 1 above)
Comment (danielhlarkin) [reply arangod/Cluster/ClusterMethods.cpp]: Do you mean returning the pair?

File: arangod/Cluster/ClusterMethods.cpp (hunk 4 above)
Comment (danielhlarkin) [reply arangod/Cluster/ClusterMethods.cpp]: I think you are correct about this. I'll figure out a way to rework this a bit.

File: arangod/Cluster/ClusterMethods.cpp (hunk 3 above)
Comment (danielhlarkin) [reply arangod/Cluster/ClusterMethods.cpp]: I suppose, yes. I chose OperationResult for consistency, but it's somewhat overkill.

File: arangod/Cluster/ClusterMethods.cpp (hunk 2 above)
Comment (graetzer) [reply arangod/Cluster/ClusterMethods.cpp]: In my view this would seem more natural
```suggestion
T val;
auto opres = handleResponsesFromAllShards(responses, 
[&val](VPackSlice slice) -> Result {
   val = doSomething(slice);
   return Result()
});
```

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/6365/

File: arangod/Cluster/ClusterMethods.cpp [hunk 5]
```diff
@@ -753,251 +839,262 @@ bool smartJoinAttributeChanged(LogicalCollection const& collection, VPackSlice c
...
-    cc->asyncRequest(coordTransactionID, "shard:" + p.first,
-                     arangodb::rest::RequestType::GET,
-                     "/_db/" + StringUtils::urlEncode(dbname) +
-                         "/_api/collection/" + StringUtils::urlEncode(p.first) +
-                         "/loadIndexesIntoMemory",
-                     std::shared_ptr<std::string const>(), headers, nullptr, 300.0);
-  }
+    // handler expects valid velocypack body (empty object minimum)
+    VPackBuffer<uint8_t> buffer;
+    VPackBuilder builder(buffer);
```
Comment (graetzer) [inline arangod/Cluster/ClusterMethods.cpp]:
```suggestion
    buffer.append(VPackSlice::emptyObjectSlice().begin(), 1)
```

File: arangod/Cluster/ClusterMethods.cpp [hunk 6]
```diff
@@ -753,251 +839,262 @@ bool smartJoinAttributeChanged(LogicalCollection const& collection, VPackSlice c
...
   for (std::pair<ShardID, std::vector<ServerID>> const& p : *shardIds) {
-    auto headers = std::make_unique<std::unordered_map<std::string, std::string>>();
-    ClusterTrxMethods::addTransactionHeader(trx, /*leader*/ p.second[0], *headers);
-    requests.emplace_back("shard:" + p.first, arangodb::rest::RequestType::GET,
-                          "/_db/" + StringUtils::urlEncode(dbname) +
-                              "/_api/collection/" +
-                              StringUtils::urlEncode(p.first) + "/count",
-                          body, std::move(headers));
+    network::Headers headers;
+    addTransactionHeaderForShard(trx, *shardIds, /*shard*/ p.first, headers);
```
Comment (graetzer) [inline arangod/Cluster/ClusterMethods.cpp]: why did you change it addTransactionHeaderForShard ?
```suggestion
    ClusterTrxMethods::addTransactionHeader(trx, /*leader*/ p.second[0], *headers);
```

Comment (danielhlarkin) [reply arangod/Cluster/ClusterMethods.cpp]: Sloppy copy/paste/edit. Will revert.

File: arangod/VocBase/Methods/Collections.h [hunk 7]
```diff
@@ -120,9 +122,10 @@ struct Collections {
       double timeout                      // single-server drop timeout
   );
 
-  static Result warmup(TRI_vocbase_t& vocbase, LogicalCollection const& coll);
+  static futures::Future<Result> warmup(TRI_vocbase_t& vocbase,
+                                        LogicalCollection const& coll);
 
-  static Result revisionId(Context& ctxt, TRI_voc_rid_t& rid);
+  static futures::Future<OperationResult> revisionId(Context& ctxt);
```
Comment (graetzer) [inline arangod/VocBase/Methods/Collections.h]: perhaps `ResultT` would be more appropriate here ? But idk

Comment (danielhlarkin) [reply arangod/VocBase/Methods/Collections.h]: Simple enough to just use the VPack in this case, and then we don't need to deal with `boost::optional`.

**Comment by danielhlarkin** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/6380/

File: arangod/Transaction/Methods.cpp [hunk 8]
```diff
@@ -2621,17 +2621,29 @@ OperationResult transaction::Methods::countCoordinatorHelper(
...
+          // reassemble counts from vpack
+          std::vector<std::pair<std::string, uint64_t>> counts;
+          TRI_ASSERT(res.slice().isArray());
+          for (VPackSlice count : VPackArrayIterator(res.slice())) {
+            TRI_ASSERT(count.isArray());
+            TRI_ASSERT(count[0].isString());
+            TRI_ASSERT(count[1].isNumber());
+            std::string key = count[0].copyString();
+            uint64_t value = count[1].getNumericValue<uint64_t>();
+            counts.emplace_back(key, value);
```
Comment (jsteemann) [inline arangod/Transaction/Methods.cpp]:
```suggestion
            counts.emplace_back(std::move(key), value);
```

**Comment by danielhlarkin** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/6395/

**Comment by danielhlarkin** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/6396/

**Comment by danielhlarkin** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/6415/

**Comment by danielhlarkin** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/6457/
